File: sensor_stack/NqDVL/src/NqDVL.cpp
#include <NqDVL/NqDVL.h>
#include <ros/ros.h>

using namespace std;
using namespace kraken_hardware;

char * pEnd;
char *b;

std::string bytes;


namespace kraken_sensors
{

  NqDVL::NqDVL(const string &name, int freq, int baudrate):SerialPort(name,baudrate),_freq(freq)
  {
    ros::NodeHandle n;
    _publisher = n.advertise<kraken_msgs::dvlData>("/kraken/dvl_data",2);
  }
  
  bool NqDVL::start(bool debug)
  {
    _debug = debug;
    openPort();
    if(isOpen())
      {
        ros::NodeHandle n;
        _timer = n.createTimer(ros::Duration(1.0/_freq),&NqDVL::timerCallBack,this);
      }
    else
      {
        return false;
      }
  }
  
  void NqDVL::timerCallBack(const ros::TimerEvent &)
  {
    readDVLData();
    getVal();
    _publisher.publish(_data);
  }
  
  void NqDVL::readDVLData()
  {
    _rawData.clear();
    std::string _garbage;
    while(ros::ok())
    {
        if(readData(_garbage, 1) != -1)
        {
            if(_garbage[0] == '$')
            {
                if(readData(_garbage, 1))
                {
                    if(_garbage[0] == '#')
                        break;
                }
            }
        }
        else
        {
            cout << "Unable to read" << endl;
        }
    }

    do{
        if(readData(_garbage, 1) != -1)
        {
            _rawData.append((_garbage));
        }
        else
        {
            break;
        }
    }while(_garbage[0] != '\r' || ros::ok());
  }

  void NqDVL::getVal()
  {
       _errorCode = strtod(b, &pEnd);
       _goodOrBad[0] = strtod(pEnd, &b);
       _goodOrBad[1] = strtod(b, &pEnd);
       _goodOrBad[2] = strtod(pEnd, &b);
       _goodOrBad[3] = strtod(b, &pEnd);
       _vAltitude[0] = strtod(pEnd, &b);
       _vAltitude[1] = strtod(b, &pEnd);
       _vAltitude[2] = strtod(pEnd, &b);
       _vAltitude[3] = strtod(b, &pEnd);
       _veloRad[0] = strtod(pEnd, &b);
       _veloRad[1] = strtod(b, &pEnd);
       _veloRad[2] = strtod(pEnd, &b);
       _veloRad[3] = strtod(b, &pEnd);
       _wveloCredit[0] = strtod(pEnd, &b);
       _wveloCredit[1] = strtod(b, &pEnd);
       _wveloCredit[2] = strtod(pEnd, &b);
       _wveloCredit[3] = strtod(b, &pEnd);
       _veloInstrument[0] = strtod(pEnd, &b);
       _veloInstrument[1] = strtod(b, &pEnd);
       _veloInstrument[2] = strtod(pEnd, &b);
       _veloInstrument[3] = strtod(b, &pEnd);
       _veloEarth[0] = strtod(pEnd, &b);
       _veloEarth[1] = strtod(b, &pEnd);
       _veloEarth[2] = strtod(pEnd, &b);
       _veloEarth[3] = strtod(b, &pEnd);
       _waterVeloInstrument[0] = strtod(pEnd, &b);
       _waterVeloInstrument[1] = strtod(b, &pEnd);
       _waterVeloInstrument[2] = strtod(pEnd, &b);
       _waterVeloInstrument[3] = strtod(b, &pEnd);
       _waterVeloEarth[0] = strtod(pEnd, &b);
       _waterVeloEarth[1] = strtod(b, &pEnd);
       _waterVeloEarth[2] = strtod(pEnd, &b);
       _waterVeloEarth[3] = strtod(b, &pEnd);
       _data.data[_dvl_roll] = _rph[0] = strtod(pEnd, &b);
       _data.data[_dvl_pitch] = _rph[1] = strtod(b, &pEnd);
       _data.data[_dvl_yaw] = _rph[2] = strtod(pEnd, &b);
       _data.data[_dvl_depth] = _altitudeEstimate = strtod(b, &pEnd);
       _data.data[_dvl_temperature] = _temperature = strtod(pEnd, &b);
       _data.data[_dvl_salinity] = _salinity = strtod(b, &pEnd);
       _data.data[_dvl_sound_speed] = _soundSpeed = strtod(pEnd, &b);
       _checksum = strtod(b, &pEnd);
  }
   
  void NqDVL::showVal()
  {  
      cout<<"starting to print the values from Errorcode.\n"<<_errorCode;
      for(int i=0; i<4; i++)
          cout<<"\n"<<"GoodOrBad["<< i<<"] : "<<_goodOrBad[i];
      for(int i=0; i<4; i++)
        cout<<"\n"<<"Valtitude["<< i<<"] : "<<_vAltitude[i];
      for(int i=0; i<4; i++)
          cout<<"\n"<<"VeloRad["<< i<<"] : "<<_veloRad[i];
      for(int i=0; i<4; i++)
          cout<<"\n"<<"WveloCredit["<< i<<"] : "<<_wveloCredit[i];
      for(int i=0; i<4; i++)
          cout<<"\n"<<"VeloInstrument["<< i<<"] : "<<_veloInstrument[i];
      for(int i=0; i<4; i++)
          cout<<"\n"<<"VeloEarth["<< i<<"] : "<<_veloEarth[i];
      for(int i=0; i<4; i++)
          cout<<"\n"<<"WaterVeloInstrument["<< i<<"] : "<<_waterVeloInstrument[i];
      for(int i=0; i<4; i++)
          cout<<"\n"<<"WaterVeloEarth["<< i<<"] : "<<_waterVeloEarth[i];
      for(int i=0; i<3; i++)
          cout<<"\n"<<"RPY["<< i<<"] : "<<_rph[i];
      cout<<"\n"<<"AltitudeEstimate : "<<_altitudeEstimate;
      cout<<"\n"<<"Temperature : "<<_temperature;
      cout<<"\n"<<"Salinity : "<<_salinity;
      cout<<"\n"<<"SoundSpeed : "<<_soundSpeed;
      cout<<"Ending at the Value of _checksum"<<_checksum;
  }

}


/*
  NqDVL::NqDVL(const std::string &name, int rate, long flag) : _dvlIO(name, rate, flag)
  {
      _dvlIO.openPort();
      if(_dvlIO.isOpen()==true)
      {
          cout<<"Port has been opened" << endl;
      }
      else
          cout<<"Port opening failed" << endl;
      cout << "BBYe" << endl;
  }
 * 
  void NqDVL::readDVLData()
  {
      _rawData.clear();
      std::string _garbage;
      while(ros::ok())
      {
          if(_dvlIO.readData(_garbage, 1) != -1)
          {
              if(_garbage[0] == '$')
              {
                  cout << "Inside"<<_garbage[0] << endl;
  
                  cout << "Inside dollar" << endl;
                  if(_dvlIO.readData(_garbage, 1))
                  {
                      if(_garbage[0] == '#')
                          break;
                  }
              }
          }
          else
          {
              cout << "Unable to read" << endl;
          }
          cout << "Inside"<<_garbage[0] << endl;
      }
  
      do{
          if(_dvlIO.readData(_garbage, 1) != -1)
          {
              _rawData.append((_garbage));
          }
          else
          {
              break;
          }
          cout << "Inside buf"<<_garbage[0] << endl;
      }while(_garbage[0] != '\r' || ros::ok());
  
      cout << _rawData.c_str() << endl;
  }
*  
  
  void NqDVL::getVal()
  {
       _errorCode = strtod(b, &pEnd);
       _goodOrBad[0] = strtod(pEnd, &b);
       _goodOrBad[1] = strtod(b, &pEnd);
       _goodOrBad[2] = strtod(pEnd, &b);
       _goodOrBad[3] = strtod(b, &pEnd);
       _vAltitude[0] = strtod(pEnd, &b);
       _vAltitude[1] = strtod(b, &pEnd);
       _vAltitude[2] = strtod(pEnd, &b);
       _vAltitude[3] = strtod(b, &pEnd);
       _veloRad[0] = strtod(pEnd, &b);
       _veloRad[1] = strtod(b, &pEnd);
       _veloRad[2] = strtod(pEnd, &b);
       _veloRad[3] = strtod(b, &pEnd);
       _wveloCredit[0] = strtod(pEnd, &b);
       _wveloCredit[1] = strtod(b, &pEnd);
       _wveloCredit[2] = strtod(pEnd, &b);
       _wveloCredit[3] = strtod(b, &pEnd);
       _veloInstrument[0] = strtod(pEnd, &b);
       _veloInstrument[1] = strtod(b, &pEnd);
       _veloInstrument[2] = strtod(pEnd, &b);
       _veloInstrument[3] = strtod(b, &pEnd);
       _veloEarth[0] = strtod(pEnd, &b);
       _veloEarth[1] = strtod(b, &pEnd);
       _veloEarth[2] = strtod(pEnd, &b);
       _veloEarth[3] = strtod(b, &pEnd);
       _waterVeloInstrument[0] = strtod(pEnd, &b);
       _waterVeloInstrument[1] = strtod(b, &pEnd);
       _waterVeloInstrument[2] = strtod(pEnd, &b);
       _waterVeloInstrument[3] = strtod(b, &pEnd);
       _waterVeloEarth[0] = strtod(pEnd, &b);
       _waterVeloEarth[1] = strtod(b, &pEnd);
       _waterVeloEarth[2] = strtod(pEnd, &b);
       _waterVeloEarth[3] = strtod(b, &pEnd);
       _rph[0] = strtod(pEnd, &b);
       _rph[1] = strtod(b, &pEnd);
       _rph[2] = strtod(pEnd, &b);
       _altitudeEstimate = strtod(b, &pEnd);
       _temperature = strtod(pEnd, &b);
       _salinity = strtod(b, &pEnd);
       _soundSpeed = strtod(pEnd, &b);
       _checksum = strtod(b, &pEnd);
  }
*/ 

File: sensor_stack/nqDVL/src/nqDVL.py
#!/usr/bin/env python

PKG = 'nqDVL'

import roslib; roslib.load_manifest(PKG)
import serial
import math
import struct
import numpy
import rospy
import sys
import os
from resources import topicHeader
from kraken_msgs.msg import dvlData
#dataList = ['error_code[0]' , 'error_code[1]' , 'error_code[2]' , 'error_code[3]' , 'good_or_bad[0]' , 'good_or_bad[1]' , 'good_or_bad[2]' , 'good_or_bad[3]' , 'v_altitude[0]' , 'v_altitude[1]' , 'v_altitude[2]' , 'v_altitude[3]' , 'velo_rad[0]' , 'velo_rad[1]' , 'velo_rad[2]' , 'velo_rad[3]' , 'wvelo_rad[0]' , 'wvelo_rad[1]' , 'wvelo_rad[2]' , 'wvelo_rad[3]' , 'wvelo_credit[0]' , 'wvelo_credit[1]' , 'wvelo_credit[2]' , 'wvelo_credit[3]' , 'velo_instrument_x' , 'velo_instrument_y' , 'velo_instrument_z' , 'velo_instrument_flag' , 'velo_earth_n' , 'velo_earth_e' , 'velo_earth_d' , 'velo_earth_flag' , 'water_velo_instrument_x' , 'water_velo_instrument_y' , 'water_velo_instrument_z' , 'water_velo_instrument_flag' , 'water_velo_earth_n' , 'water_velo_earth_e' , 'water_velo_earth_d' , 'water_velo_earth_flag' , 'roll' , 'pitch' , 'heading' , 'altitude_estimate' , 'temperature' , 'salinity' , 'sound-speed' , 'check-sum',]

dataList = ['pitch' , 'roll', 'yaw', 'timeStamp', 'temp', 'depth', 'soundVel', 'BIT', 'transVelCurr', 'longVelCurr', 'normVelCurr', 'NA1', 'statusCurr', 'transVel', 'longVel', 'normVel', 'NA2', 'status', 'garbage', 'eastDist' , 'northDist', 'upDist', 'altitude', 'timedata']
#BIT : built-in-test record
#transVelCurr : tranverse velocity relative to current
#longVelCurr : longitudinal velocity relative to current
#normVelCurr : normal velocity relative to current
#NA1 : this data is not used
#statusCurr : 'A' => Valid data , 'I' => Invalid data
#transVel : transverse velocity
#longVel : longitudinal velocity
#normVel : normal velocity
#NA2 : this data is not used
#status : 'A' => Valid data , 'I' => Invalid data
#eastDist : listed as 'Not Used' in application note
#northDist :  listed as 'Not Used' in application note
#upDist :  listed as 'Not Used' in application note
#altitude : DVL altitude, '9999.99' means invalid altitude information
#timedata :  listed as 'Not Used' in application note

pub = rospy.Publisher(topicHeader.SENSOR_DVL, dvlData, queue_size = 2)
rospy.init_node('dvldata', anonymous=True)

## Code to find port automatically
find = os.popen('dmesg | grep pl2303')
port = find.read()
num = port.split('now attached to ')
portName = '/dev/'+(num[1].split('\n'))[0]
##
'''
if (len(sys.argv) == 2):
	num = str(sys.argv[1])
else:
	num = '1'
'''
dvl = serial.Serial(portName, 9600)

rawData = ''
pubData =  dvlData()

## DVL config
dvl.stopbits = 2
##


def readDVLdata():
	global rawData
	garbage = ''
	rawData = ''
	
	while not rospy.is_shutdown():
		garbage = dvl.read(1)
		
		if (garbage != -1):
			
			if (garbage[0] == '$'):
				garbage = dvl.read(1)
				
				if (garbage[0] == '#'):
					break
				
		else:
			print 'Unable to read'
			return
				
	while True :
		rawdata = '$#'
		garbage = dvl.read(1)
		
		if (garbage != -1):
			rawData += garbage
			

		
		if (garbage[0] == '$'):
			rawData += garbage
			while(garbage[0] != '\r'):
				garbage = dvl.read(1)
				rawdata += garbage
			break

	
def getVal(rawdata):
	global pubData
	floatData = []
	data = rawdata.split('\r')
	try:
		for i in range(1,len(data)-1):
			datal = (data[i]).split(',')
		
			for j in range(1,len(datal)):
				if (datal[j] == 'A'):
					floatData += [1.0]
				elif(datal[j] == 'I'):
					floatData += [0.0]
				else:
					floatData += [float(datal[j])]

		print floatData,len(floatData)
		
		if(len(floatData) == 24):       
			
		#print (floatData)

			pubData.data[0] = floatData[1]   #roll
			pubData.data[1] = floatData[0]   #pitch
			pubData.data[2] = floatData[2]   #yaw
			pubData.data[3] = -floatData[15]/100.0  #Longitudinal Vel
			pubData.data[4] = floatData[14]/100.0  #Y Vel
			pubData.data[5] = floatData[16]/100.0  #Normal Vel
			pubData.data[6] = floatData[22]  #altitude
			pubData.data[7] = floatData[5]   #depth  (does not work in our model)
			pubData.data[8] = floatData[4]   #temperature
			pubData.data[9] = floatData[7]
	except Exception, e:
		pass
	   #Sound Velocity
		
if __name__ == '__main__':
	

	if (not dvl.isOpen):
		dvl.open()

	if (dvl.isOpen) :
		print 'Serial port opened successfully'
	else:
		print 'Error in opening port'
#    while True :
		
#        k = ''
#        while(k != '\r'):
#            k = dvl.read(1)
#            if (k == '$'):
#                print 'got it'
#            print k
#        print 'out'
	
	r = rospy.Rate(10)
	count = 1
#        k = ''    
	while not rospy.is_shutdown():
	   
		readDVLdata()
		#print 'data count : ',count 
		getVal(rawData)
		#print ''
		#count += 1
		
		#for i,a in enumerate(dataList):
		#    print i,' : ',a
		#rospy.loginfo(pub)
		pub.publish(pubData)
		
	   
		r.sleep()
		
	
	
	dvl.close()
